### ISSUE TITLE
Uncaught OutOfBoundsException: Required package "vimeo/psalm" is not installed: cannot detect its version

### ISSUE THREAD

**Issue by Aerendir** [vimeo/psalm#3774]
I'm running Psalm in CI (GitHub Actions) and installing composer dependencies at lowest possible levels, I get this error:

    composer update --prefer-stable --prefer-lowest --no-interaction --no-suggest

Trying to reproduce the issue locally, I found that the problem is in the lib `muglug/PackageVersions`.

Its lowest version, in fact, is `1.2.0`, but installing this version causes the error reported.

# Suggested fix

In `composer.json´, require minimum version of `muglug/PackageVersions`:

    "muglug/package-versions-56": "^1.2.4",

# Workaround

Require directly in your `composer.json` the indicated version:

    composer req muglug/package-versions-56:^1.2.4 --dev

# Context

I tried also older versions and it turned out that the minimum version that solves the issue is the `1.2.4`.

To check the real problem, you can check the failed job in GitHub Actions:

https://github.com/Aerendir/PHPArrayWriter/runs/849592095?check_suite_focus=true#step:6:4

**Comment by muglug** [comment]
`muglug/package-versions-56` was removed from Psalm when it shifted to PHP 7.1, many moons ago. You should not need it in any guise.

**Comment by Aerendir** [comment]
Please, give a look at the linked action in GitHub: you’ll see that psalm requires it somewhere.

**Comment by Aerendir** [comment]
The point is this: the error happens and from somewhere it has to come.

Setting the dependency as indicated, solved the problem.

I’ll look again at the code anyway, but, please, do the same if you can.

**Comment by orklah** [comment]
Your version of phpunit plugin for Psalm is too old.

It uses version 0.5.0 but this was changed in 0.5.7:
https://github.com/psalm/psalm-plugin-phpunit/commit/82ef477f17b95e2be5a776ed9e22afcb59015c98#diff-b5d0ee8c97c7abd7e3fa29b9a27d1780

**Comment by weirdan** [comment]
And current version is something like 0.10.x, so definitely worth updating.

**Comment by Aerendir** [comment]
Many thank you for your comments :)

I'm using `@stable` for `psalm/plugin-phpunit`, so I don't have to deal with versions: I simply set the version of Psalm, and the rest is done by Composer.

But it seems this strategy is not working...

Do you think at any solution that doesn't force me to set explicitly a version for `psalm/plugin-phpunit`?

**Comment by orklah** [comment]
@Aerendir Can you explain what you're trying to gain by launching an old version of a tool first?

You're basically running the same thing but with more bugs and less features

**Comment by Aerendir** [comment]
@orklah , I want to test the code on various versions of PHP and with various versions of the same dependencies.

As those are libraries, I don't want to force the end developer to stick to a specific version of my dependencies.

So I test my library with the highest possible dependencies and with the lowest possible dependencies.

For example, I use some Symfony components: I test my libraries with version 3, 4 and 5 to ensure maximum interoperability.

In doing this, I have to set wide versions for all other dev tools (like Psalm), so Composer can install what can be installed depending on the dependencies specified by the library that uses my library.

If I make the versions fixed, I force the developer to be compatible with my, while it should be the exact opposite: is my library that has to leave free the developer of installing what (s)he likes.

For this reason in some jobs, the versions installed are very old and happens that the library `muglug/package-versions-56` is installed.

**Comment by orklah** [comment]
I fully understand why you would want to test Your code and it's dependancies in every version you support and it's a good approach.

But for dev dependancies and furthermore for tools, it's really weird. Nobody is supposed to run phpcs on your code with PHP7.1 and an old version of phpcs. The same goes for Psalm.

If you're worried about conflict on dev dependancies, I'd argue that you should not include psalm on your composer.json at all.

Running `composer require --dev vimeo/psalm:version` directly in your CI will avoid the problem altogether.

If you want to support old version of every tool you use, you'll have to micromanage every version of every tool that has a bug you don't want or that don't have a feature you need. And that's what happens here.

**Comment by weirdan** [comment]
> If you're worried about conflict on dev dependancies, I'd argue that you should not include psalm on your composer.json at all.

Dev dependencies do not conflict in composer, as the only dev dependencies that are ever installed are for the root package, only. Runtime dependencies of your dev-deps may conflict with other packages, so your suggestion still has sense, but see below.

> Running composer require --dev vimeo/psalm:version directly in your CI will avoid the problem altogether.

This may still introduce conflicts. E.g psalm depends on `symfony/console` of a certain minimum version, but your lib/app is compatible with older version of that. Installing `vimeo/psalm` (even as a dev dependency) will bump `symfony/console` as well - thus you would no longer be testing with the minimum supported version of `symfony/console`. For this reason I would suggest to use `bamarni/composer-bin-plugin` instead.

**Comment by orklah** [comment]
I was thinking about conflicts with dependancies used by the lib user, but I forgot that dev dependancies are only installed for the root package :)

So it makes even less sense running CI with old versions of dev tools.

**Comment by Aerendir** [comment]
Ok, this is becoming a good discussion ❤️

So, first of all, let me explain better that I don’t use all the dev tool with the matrix of versions, but only the static analytics tools like Psalm, Phan and PHPStan.

This is because I don’t have time to reach 100% code coverage with unit tests, so I rely on static analysis to spot some possible bugs.

With the matrix of versions I also use PHPUnit, that is another level of check, but, as I said, I don’t reach 100% and sometimes there are no tests at all.

That said, the solutions I tried to avoid conflicts are basically two:

1. Using Phars (through the very great Phive tool)
2. Bamarni/composer-bin-plugin

The first approach (phars) is not practicable as the plugins of static analysis tools are not compatible with such versions as, on one hand, the phars often have the names of classes randomized and also because often (practically ever) the plugins works only when installed together with the tools in the same vendor folder.
So this approach may work for something like phpcs, that hasn’t plugins, but cannot work with tools like Phan, phpstan or psalm (about psalm I think, but I’m not sure as I abandoned the approach before trying it with Psalm).

The other option is bamarni/composer-non-plugin.

This seems to be a good solution, but it has a major problem: it seems not compatible with Dependabot.

And Dependabot is really a great ally, as it does all the dirty job for us.

But I saw that in the version 2 developed by GitHub, it is possible to specify various composer.lock (forgive me if I’m terminologically incorrect, but I don’t remember how GitHub’s Dependabot names them).

This may mean that I can use bamarni solution and then specify each dependency in the Dependabot.yml file to be updated separately.

But this is a way I have to explore further.

The approach suggested, of installing the dependency directly in the CI snc, as well explained, is not practicable as it doesn’t plow an (almost) rigorous control on the versions.

So, the best bet would be bamarni and this is exactly what I will try.
